### PR TITLE
Antagonist uplinks no longer depend on gamemode

### LINKED ISSUE #5603: Late-joining traitors during spy thief rounds have a restricted buy pool
[BUG]

**Describe the bug**
If you latejoin a spy round and roll traitor, your uplink purchases are restricted to things that spies can find (anything without `datum/game_mode/spy_theft` in `blockedmode`.

**To Reproduce**
1. Latejoin as a traitor on a spy round
2. Can't buy the c saber, shotgun, power gloves, etc.

**Expected behavior**
Traitors have access to their full buylist.

**Additional context**
Haven't seen this myself; reported on discord a few days ago, and makes sense based on the way that `blockedmode` works.

### PR DESCRIPTION
**Pull request by Zonespace27**
[QoL]
## About the PR 
This PR is a backend partial rework to uplinks, making it so that uplinks depend on bitflags on the buylist datum, and vice versa, to determine what can and cannot be bought, versus being based on the gamemode.
Fixes #5603

## Why's this needed? 
This is better than the hacky solution of before, and I may have plans

### CODE REVIEW

File: code/obj/item/uplinks.dm [hunk 1]
```diff
@@ -995,7 +1033,7 @@ Note: Add new traitor items to syndicate_buylist.dm, not here.
 
 	proc/setup()
 		for (var/datum/syndicate_buylist/S in syndi_buylist_cache)
-			if (istype(S, /datum/syndicate_buylist/commander) && !src.commander_buylist.Find(S))
+			if ((istype(S, /datum/syndicate_buylist/commander) || (S.can_buy & purchase_flags)) && !src.commander_buylist.Find(S))
```
Comment (ZeWaka) [inline code/obj/item/uplinks.dm:1036]:
```suggestion
			if ((istype(S, /datum/syndicate_buylist/commander) || (S.can_buy & purchase_flags)) && !(S in src.commander_buylist))
```

File: code/datums/syndicate_buylist.dm [hunk 2]
```diff
@@ -1136,56 +1107,57 @@ This is basically useless for anyone but miners.
...
 	desc = "Three deadly throwing darts that embed themselves into your target."
 	job = list("Assistant","Technical Assistant","Medical Assistant","Staff Assistant","Bartender","Clown")
-	blockedmode = list(/datum/game_mode/revolution)
+	can_buy = UPLINK_TRAITOR | UPLINK_SPY_THIEF | UPLINK_SPY
 
 // round specific
 
-/datum/syndicate_buylist/generic/revflash
+/datum/syndicate_buylist/generic/head_rev
+	name = "Head Rev Buylist Parent"
```
Comment (ZeWaka) [inline code/datums/syndicate_buylist.dm:1123]: then make it abstract

File: code/datums/syndicate_buylist.dm [hunk 3]
```diff
@@ -1038,8 +1004,8 @@ This is basically useless for anyone but miners.
 	name = "You shouldn't see me!"
```
Comment (ZeWaka) [inline code/datums/syndicate_buylist.dm:1004]: abstract

File: code/datums/syndicate_buylist.dm [hunk 4]
```diff
@@ -65,156 +64,143 @@ proc/build_syndi_buylist_cache()
...
 	cost = 0
 	desc = "You shouldn't see me!"
```
Comment (ZeWaka) [inline code/datums/syndicate_buylist.dm:66]: abstract

**Review by ZeWaka — CHANGES_REQUESTED** [review]

File: code/datums/syndicate_buylist.dm (hunk 2 above)
Comment (Zonespace27) [reply code/datums/syndicate_buylist.dm:1128]: fixed for all cases in the buylist datums

**Pull request merged**